File: 3719-longest-balanced-subarray-i/3719-longest-balanced-subarray-i.py
class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        ans = 0
        for i in range(0, len(nums)):
            even, odd = set(), set()
            for j in range(i, len(nums)):
                if nums[j] % 2 == 0:
                    even.add(nums[j])
                else:
                    odd.add(nums[j])
                
                if len(even) == len(odd): ans = max(ans, j - i + 1)
        
        return ans

# array, hash map
# time: O(n^2)
# space: O(n)

File: 3719-longest-balanced-subarray-i/3719-longest-balanced-subarray-i.go
func longestBalanced(nums []int) int {
	ans := 0
	for i := 0; i < len(nums); i++ {
		evenSet, oddSet := map[int]bool{}, map[int]bool{}
		for j := i; j < len(nums); j++ {
			if nums[j]%2 == 0 { // even
				evenSet[nums[j]] = true
			} else { // odd
				oddSet[nums[j]] = false
			}

			if len(evenSet) == len(oddSet) {
				ans = max(ans, j-i+1)
			}
		}
	}

	return ans
}

// array, hash map
// time: O(n^2)
// space: O(n)